salt/modules/virt.py

```diff
@@ -2500,16 +2501,533 @@ def revert_snapshot(name, vm_snapshot=None, cleanup=False, **kwargs):
+
+        elif child.tagName == "migration_features":
+            migration = {
+                'live': bool(child.getElementsByTagName('live')),
+                'transports': []
+            }
+            for transport in child.getElementsByTagName('uri_transport'):
+                migration['transports'].append(_get_xml_element_text(transport))
+            result['migration'] = migration
+
+        elif child.tagName == "topology":
+            cells = []
+            for cell in child.getElementsByTagName('cell'):
+                cells.append(_parse_caps_cell(cell))
+            result['topology'] = {'cells': cells}
+
+        elif child.tagName == 'cache':
+            banks = []
+            for bank in child.getElementsByTagName('bank'):
+                banks.append(_parse_caps_bank(bank))
+            result['cache'] = {'banks': banks}
+
+    secmodels = []
+    for secmodel in host.getElementsByTagName('secmodel'):
+        model = {
+            'model': _get_xml_child_text(secmodel, 'model', None),
+            'doi': _get_xml_child_text(secmodel, 'doi', None)
+        }
+        labels = []
+        for label in secmodel.getElementsByTagName('baselabel'):
+            labels.append({
+                'type': label.getAttribute('type'),
+                'label': _get_xml_element_text(label)
+            })
+        if labels:
+            model['baselabels'] = labels
+        secmodels.append(model)
+    if secmodels:
+        result['security'] = secmodels
+
+    return result
+
+
+def capabilities(**kwargs):
+    '''
+    Return the hypervisor connection capabilities.
+
+    ..versionadded:: Fluorine
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' virt.capabilities
+    '''
+    conn = __get_conn(**kwargs)
+    caps = _capabilities(conn)
+    conn.close()
+
+    result = {
+        'host': _parse_caps_host(caps.getElementsByTagName('host')[0]),
+        'guests': []
+    }
+
+    for guest in caps.getElementsByTagName('guest'):
+        result['guests'].append(_parse_caps_guest(guest))
+
+    return result
+
+
+def _parse_caps_supported(node):
+    '''
+    Return a boolean value for a node with a supported attribute.
+    None is returned if there is no 'supported' attribute or if its
+    value is not one of 'yes' or 'no'.
+    '''
+    supported = node.getAttribute('supported')
+    if not supported:
+        return None
+    if supported == 'yes':
+        return True
+    elif supported == 'no':
+        return False
+    return None
+
+
+def _parse_caps_enum(node):
+    '''
+    Return a tuple containing the name of the enum and the possible values, or None
+    if the node doesn't fit the expected format.
+    '''
+    name = node.getAttribute('name')
+    if not name:
+        return None
+
+    values = []
+    for value in node.getElementsByTagName('value'):
+        values.append(_get_xml_element_text(value))
+    return (name, values)
```

Comment:
This is why SESE matters. You return different types here. Once just `None`, other case tuple. That said, you cannot anymore do something like:

```python
name, values = _parse_caps_enum(node)
```

Once you did, it will crash, in case node has no attribute `name`. So I would write it this way:

```python
def _parse_caps_enum(node):
return (node.getAttribute('name') or None,
        [_get_xml_element_text(value) for value in node.getElementsByTagName('value')])